File: expofirst/modules/kmp-bridge/src/KmpBridgeModule.ts
import { NativeModule, requireNativeModule } from 'expo';

declare class KmpBridgeModule extends NativeModule<{}> {
  /** Synchronous call into the KMP shared module. */
  greet(name: string): string;
  /** Async call into the KMP shared module, resolved via a Promise. */
  greetAsync(name: string): Promise<string>;
}

export default requireNativeModule<KmpBridgeModule>('KmpBridge');

File: expofirst/modules/kmp-bridge/ios/KmpBridgeModule.swift
import ExpoModulesCore
import Shared

public class KmpBridgeModule: Module {
  // Single instance of the shared KMP logic, reused across calls.
  private let greeting = Greeting()

  public func definition() -> ModuleDefinition {
    Name("KmpBridge")

    // Synchronous call straight into the KMP `commonMain` logic.
    Function("greet") { (name: String) in
      return self.greeting.greet(name: name)
    }

    // Async variant — same KMP call, returned via a Promise on the JS side.
    AsyncFunction("greetAsync") { (name: String) in
      return self.greeting.greet(name: name)
    }
  }
}
